### ISSUE TITLE
NextJS redirect next/navigation 13.4.2

### ISSUE THREAD

**Issue by MohammedQureshi** [vercel/next.js#49964]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Doc https://nextjs.org/docs/app/api-reference/functions/redirect

Example code 
if (isLoggedIn) {
  redirect('/feed');
}

When redirect is called it returns the error below 

Error: NEXT_REDIRECT
    at getRedirectError (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:40:19)
    at redirect (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:46:11)
    at Login (webpack-internal:///(sc_server)/./src/app/login/page.tsx:32:74)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: 'NEXT_REDIRECT;replace;/'
}
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue

N/A

### To Reproduce

Add redirect('/page') <- Into any class and run

### Describe the Bug

Does not redirect and throws error below

Error: NEXT_REDIRECT
    at getRedirectError (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:40:19)
    at redirect (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:46:11)
    at Login (webpack-internal:///(sc_server)/./src/app/login/page.tsx:32:74)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: 'NEXT_REDIRECT;replace;/'
}

### Expected Behavior

To redirect onto correct page 

### Which browser are you using? (if relevant)

Chrome 113.0.5672.92 

### How are you deploying your application? (if relevant)

yarn dev

**Comment by bernatfortet** [comment]
I was having this issue and I solved it by properly using await on some methods I was waiting on. I hope that helps.

**Comment by timneutkens** [comment]
> Add redirect('/page') <- Into any class and run

This is not a great reproduction. Please check the issue template, we provide many steps you can take to provide useful reproduction steps.

**Comment by MohammedQureshi** [comment]
`import { cookies } from 'next/headers';
import { redirect } from 'next/navigation';

export default async function RestrictedPage({ children }: { children: React.ReactNode }) {

    const cookieStore = cookies();
    const token = cookieStore.get('token');

    if(token) {
        await fetch('http://localhost:8443/api/1.0/user', {
            method: 'GET',
            headers: {
                'Accept': '*/*',
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${token.value}`
            },
        })
        .then(response => {
            redirect('/login')
        })  
        .catch(err => console.log(err + "ERROR"))
    }

    return (
        <div>
            {children}
        </div>
    )
}`


Class Name RestrictedPage.tsx

Error
Error: NEXT_REDIRECTERROR

**Comment by MohammedQureshi** [comment]
`import { cookies } from 'next/headers';
import { redirect } from 'next/navigation';

async function checkToken() {
    const cookieStore = cookies();
    const token = cookieStore.get('token');

    if(token) {
        await fetch('http://localhost:8443/api/1.0/user', {
            method: 'GET',
            headers: {
                'Accept': '*/*',
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${token.value}`
            },
        })
        .then(response => {
            redirect('/login')
        })
    }
}

export default function RestrictedPage({ children }: { children: React.ReactNode }) {
    checkToken()
    return (
        <div>
            {children}
        </div>
    )
}`


ERRRR:

- error unhandledRejection: Error: NEXT_REDIRECT
    at getRedirectError (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:40:19)
    at redirect (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:50:11)
    at eval (webpack-internal:///(sc_server)/./src/app/components/RestrictedPage.tsx:26:70)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async checkToken (webpack-internal:///(sc_server)/./src/app/components/RestrictedPage.tsx:18:9) {
  digest: 'NEXT_REDIRECT;replace;/login'

**Comment by MohammedQureshi** [comment]
<img width="674" alt="image" src="https://github.com/vercel/next.js/assets/34938712/018a3f2f-29b5-4844-8203-a89b6c74a6a3">

^ More of a clear view

**Comment by julianpoma** [comment]
Not working for me either. Basically the error happens if you have an async `GET` in my case.

```js
export async function GET() {
  const isAuthenticated = await getSession()

  if (!isAuthenticated) {
    redirect('/login')
  }

  redirect('/dashboard')
} 
```

This might be hacky, but its working on my end.

```js
export async function GET(request) {
  const isAuthenticated = await getSession()
  const url = new URL(request.url)

  if (!isAuthenticated) {
    url.pathname = '/login'
    return NextResponse.redirect(url)
  }

  url.pathname = '/dashboard'
  return NextResponse.redirect(url)
} 
```

**Comment by balazsorban44** [comment]
This issue has been automatically closed because it received no activity for a month and had no reproduction to investigate. If you think it was closed by accident, please leave a comment. If you are running into a similar issue, please open a new issue with a reproduction. Thank you.

**Comment by timneutkens** [comment]
<img width="296" alt="CleanShot 2023-07-04 at 12 14 21@2x" src="https://github.com/vercel/next.js/assets/6324199/60ae0c8a-76dc-4201-b3d2-d9cd69229555">

You forgot to await checkToken, so it runs out of bound of the function. Changing it to be `async function` and then adding `await` likely fixes your issue.

@julianpoma can you open a new issue? Looks unrelated to the one reported. Thanks!
